middleware/types.go

```diff
@@ -30,10 +32,17 @@ type CustomResource interface {
 	Delete(context.Context, p.DeleteRequest) error
 }
 
+// ConstructRequest contains request details based on Pulumi Go SDK types.
+type ConstructRequest struct {
+	p.ConstructRequest
+	Inputs  comProvider.ConstructInputs
+	Options pulumi.ResourceOption
+}
+
 // ComponentResource provides a shared definition of a Pulumi component resource for
 // middleware to use.
 type ComponentResource interface {
-	Construct(context.Context, p.ConstructRequest) (p.ConstructResponse, error)
+	Construct(*pulumi.Context, ConstructRequest) (pulumi.ComponentResource, error)
```

Comment:
@iwahbe would you prefer that the transition into a program context happen outside of, or within, this `Construct` method.  That is, in `dispatch` or in `infer`. I know you mentioned a preference elsewhere and it would be an easy adjustment to make here.

Reply:
I'd like all the languages to be consistent if we can. Currently the other languages transition to a program context before calling the users Construct code, even for a non-infer'd provider. So I think we should either do that here, or add something to the other languages to go from "raw mode" to "program mode" like the Host interface seems to be doing in this change.

Reply:
I'd prefer that the construct function had the same signature as in the top level provider (just like `CostomResource`. `middleware` types exist to provide a resource specific abstraction, *not* a higher level abstraction.
```suggestion
	Construct(context.Context p.ConstructRequest) (p.ConstructResponse, error)
```

Reply:
(we discussed offline how the go provider cannot eagerly transition into a program context)

I took Ian's suggestion to shift the transition into infer itself, to keep dispatch decoupled from the decision to enter a program context:
https://github.com/pulumi/pulumi-go-provider/pull/338/commits/820a2bcb0e5f050c5e3dc43fa70e72545ffa639f